PowerPointLabs/PowerPointLabs/SyncLab/SyncLabShapeStorage.cs

```diff
@@ -30,22 +39,44 @@ private SyncLabShapeStorage() : base()
             ClearShapes();
         }
 
-        // Saves shape in storage
-        // Returns a key to find the shape by,
-        // or null if the shape cannot be copied
-        public string CopyShape(Shape shape)
+        public Shapes GetTemplateShapes()
+        {
+            return Slides[FormatStorageSlide].Shapes;
+        }
+
+        /// <summary>
+        /// Saves shape in storage
+        /// Returns a key to find the shape by,
+        /// or null if the shape cannot be copied
+        /// </summary>
+        /// <param name="shape"></param>
+        /// <param name="formats">Required for msoPlaceholder</param>
+        /// <returns>identifier of copied shape</returns>
+        public string CopyShape(Shape shape, Format[] formats)
         {
-            // copies a shape, and returns a shape name
-            shape.Copy();
             Shape copiedShape = null;
-            try
+            if (shape.Type == MsoShapeType.msoPlaceholder)
             {
-                copiedShape = Slides[0].Shapes.Paste()[1];
+                copiedShape = ShapeUtil.CopyMsoPlaceHolder(formats, shape, GetTemplateShapes());
             }
-            catch
+            else
+            {
+                try
+                {
+                    shape.Copy();
+                    copiedShape = Slides[0].Shapes.Paste()[1];
+                }
+                catch
+                {
+                    copiedShape = null;
+                }
+            }
+
+            if (copiedShape == null)
```

Comment:
Why do you catch and set copiedShape to null, then proceed to check if copiedShape is null and return null?

Reply:
The if block that comes before might set copiedShape to null.